commit 3269740c41773424b0dbff3eb93ba92a8ee34782 (refs/changes/68/3868/2)
Author: Marsell Kukuljevic <marsell@joyent.com>
Date:   2018-04-25T17:40:23+00:00 (1 year, 5 months ago)
    
    TRITON-312 Extend plugin-manager so updating NICs is possible
    Reviewed by: Pedro P. Candel <pedro@joyent.com>
    Approved by: Pedro P. Candel <pedro@joyent.com>

diff --git a/lib/backends/sdc/networks.js b/lib/backends/sdc/networks.js
index 40e6533..74ec05a 100644
--- a/lib/backends/sdc/networks.js
+++ b/lib/backends/sdc/networks.js
@@ -798,6 +798,14 @@ function addNetworksToContainerPayload(opts, container, payload, callback) {
                 return;
             }
             externalNetworkByName(opts, container, payload, next);
+        },
+
+        function runModifyProvisionNetworksPlugins(_, next) {
+            opts.app.plugins.modifyProvisionNetworks({
+                account: opts.account,
+                networks: payload.networks,
+                req_id: opts.req_id
+            }, next);
         }
     ]}, function (err) {
         if (!err) {
diff --git a/lib/plugin-manager.js b/lib/plugin-manager.js
index d6cce99..7496f90 100644
--- a/lib/plugin-manager.js
+++ b/lib/plugin-manager.js
@@ -171,7 +171,7 @@ function init(app) {
             }
 
             assert.notEqual(supportedPluginHooks.indexOf(apiName), -1,
-                'supportedPluginFunctions[apiName]');
+                'supportedPluginFunctions["' + apiName + ']"');
 
             var initedPlugin = plugin[apiName](self.api, pluginCfg);
             self.hooks[apiName].push(initedPlugin);
@@ -305,3 +305,37 @@ function postProvision(opts, cb) {
 
     callfuncs();
 };
+
+
+/*
+ * This hook is run during provisioning, just before the creation of a
+ * container. It performs no checks, and returns no error. It modifies the
+ * opts.networks argument.
+ */
+PluginManager.prototype.modifyProvisionNetworks =
+function modifyProvisionNetworks(opts, cb) {
+    assert.object(opts, 'opts');
+    assert.object(opts.account, 'opts.account');
+    assert.arrayOfObject(opts.networks, 'opts.networks');
+    assert.uuid(opts.req_id, 'opts.req_id');
+    assert.func(cb, 'cb');
+
+    var hooks = this.hooks.modifyProvisionNetworks;
+    var funcs = hooks.map(function wrapFunc(func) {
+        return function (_, next) {
+            func(opts, next);
+        };
+    });
+
+    // Runs every plugin (if any).
+    function callfuncs() {
+        if (funcs.length === 0) {
+            return cb();
+        }
+
+        var func = funcs.pop();
+        return func(opts, callfuncs);
+    }
+
+    callfuncs();
+};
diff --git a/package.json b/package.json
index b3781e9..65e19a9 100644
--- a/package.json
+++ b/package.json
@@ -1,6 +1,6 @@
 {
   "name": "sdc-docker",
-  "version": "0.5.1",
+  "version": "0.5.2",
   "author": "Joyent (joyent.com)",
   "private": true,
   "dependencies": {
